dns_server/src/crawler/core/mod.rs

```diff
@@ -0,0 +1,589 @@
+        types::{ConnectivityEvent, PeerInfo},
+        NetworkingService,
+    },
+    peer_manager::global_ip::IsGlobalIp,
+    types::peer_address::PeerAddress,
+};
+
+/// How many outbound connection attempts can be made per heartbeat
+const MAX_CONNECTS_PER_HEARTBEAT: usize = 25;
+
+/// When the server drops the unreachable node address. Used for negative caching.
+const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+pub const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressTransition {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+    },
+    Connecting {
+        /// Same as in the Disconnected state
+        fail_count: u32,
+    },
+    Connected {},
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: Duration,
+
+    /// Whether the address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+
+    /// Whether the address was added from the command line
+    user_added: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerCoreConfig {
+    /// Default p2p port on which nodes normally accept inbound connections
+    pub default_p2p_port: u16,
+}
+
+/// The `CrawlerCore` is the component that communicates with Mintlayer peers using p2p,
+/// and based on the results, commands the DNS server to add/remove ip addresses.
+/// The `CrawlerCore` emits events that communicate whether addresses were reached or,
+/// are unreachable anymore.
+pub struct CrawlerCore<N: NetworkingService, C> {
+    /// Time of some monotonic timer, started from 0
+    now: Duration,
+
+    /// Crawler config
+    config: CrawlerCoreConfig,
+
+    /// Chain config
+    chain_config: Arc<ChainConfig>,
+
+    /// Map of all known addresses (including currently unreachable)
+    addresses: BTreeMap<N::Address, AddressData>,
+
+    /// Map of all currently connected outbound peers
+    outbound_peers: BTreeMap<N::PeerId, N::Address>,
+
+    /// CoreCallback implementation
+    _phantom: PhantomData<C>,
+}
+
+pub enum CoreEvent<N: NetworkingService> {
+    Timer { period: Duration },
+    Connectivity { event: ConnectivityEvent<N> },
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum CoreCommand<N: NetworkingService> {
```

Comment:
As discussed, please rename to `CrawlerP2pEvent`, which should emit events about addresses, without concrete decisions on what to do with addresses that are reachable.

Reply:
> which should emit events about addresses, without concrete decisions on what to do with addresses that are reachable.

I tried that, but it did not work. Some addresses should be kept in the DB even if they are temporarily unreachable. This is needed if the server is restarted while having temporary network problems. This could be solved by having a third state (temporarily unreachable), but this would complicate the logic in `CrawlerManager`, which I prefer not to do. Also later we may want to store some additional details in the DB (for example `fail_count`).
Considering all this, I think we should keep what we have now.

Reply:
In its current form it's simply wrong and unnecessarily complex. Having a mutable state in a random callback doing everything is very difficult to understand. It's not imperative not functional programming. We should strive to do things right. I suggest you try to do the right thing (whether it requires the new state or not), and then we try to improve it. This is much better than giving up this early. 

Again, what used to be called "core" doesn't have to know that there's a database or DNS server.

Reply:
OK, I updated again. Now `Crawler` only notifies about address state changes:
```
pub trait Callback<N: NetworkingService> {
...
    fn address_updated(
        &mut self,
        address: &N::Address,
        old_state: &AddressState,
        new_state: &AddressState,
    );
}
```
And `CrawlerManager` now reacts as needed (updates DNS or DB).
Old address state is passed to keep `CrawlerManager` stateless:
```
        match (old_state.is_persistent(), new_state.is_persistent()) {
            (false, true) => {
                storage::update_db(&self.storage, |tx| tx.add_address(&address.to_string()))
                    .expect("update_db must succeed (add_address)");
            }
            (true, false) => {
                storage::update_db(&self.storage, |tx| tx.del_address(&address.to_string()))
                    .expect("update_db must succeed (del_address)");
            }
            _ => {}
        }
```